### PR TITLE
Texture: add clamp and mirror wrapping modes

### PR DESCRIPTION
**Pull request by njroussel**
This PR adds a `WrapMode` `enum` to extend the possible texture wrapping modes with: clamping, repeating.

The `WrapMode` is used both by the `eval()` (`eval_cuda()` and `eval_enoki()`) method with the existing filtering methods, as well as all the cubic spline interpolation methods (`eval_cubic()`, `eval_cubic_helper`, `eval_cubic_grad`).

Tests cover the new wrapping modes in all aforementioned methods, for 1, 2 and 3 dimensional textures.

Notes:
I do not know the fine details of the current & future cubic spline interpolation methods use cases, but it seems to me like we could expose them through an unified interface by adding a `CubicSpline` variant to `FilterMode`. Alternatively we maybe should/could add a warning when using the cubic spline interpolation methods with the `FilterMode` being set to `Nearest`.

### CODE REVIEW

File: src/python/texture.h [hunk 1]
```diff
@@ -7,19 +10,24 @@ void bind_texture(py::module &m, const char *name) {
...
-                 return new Tex(shape.data(), channels, migrate, filter_mode);
+                         ek::FilterMode filter_mode, ek::WrapMode wrap_mode) {
+                 return new Tex(shape.data(), channels, migrate, filter_mode,
+                                wrap_mode);
              }),
              "shape"_a, "channels"_a, "migrate"_a = true,
-             "filter_mode"_a = ek::FilterMode::Linear)
-        .def(py::init<const typename Tex::TensorXf &, bool, ek::FilterMode>(),
+             "filter_mode"_a = ek::FilterMode::Linear,
+             "wrap_mode"_a   = ek::WrapMode::Repeat)
```
Comment (Speierers) [inline src/python/texture.h:19]:
```suggestion
             "wrap_mode"_a = ek::WrapMode::Repeat)
```

File: src/python/texture.h [hunk 2]
```diff
@@ -7,19 +10,24 @@ void bind_texture(py::module &m, const char *name) {
...
+        .def(py::init<const typename Tex::TensorXf &, bool, ek::FilterMode,
+                      ek::WrapMode>(),
              "tensor"_a, "migrate"_a = true,
-             "filter_mode"_a = ek::FilterMode::Linear)
+             "filter_mode"_a = ek::FilterMode::Linear,
+             "wrap_mode"_a   = ek::WrapMode::Repeat)
```
Comment (Speierers) [inline src/python/texture.h:24]:
```suggestion
             "wrap_mode"_a = ek::WrapMode::Repeat)
```

File: tests/texture.cpp [hunk 3]
```diff
@@ -16,57 +16,94 @@ using ArrayD2f = ek::Array<DFloat, 2>;
...
-
-            Array1f pos(rng_2.next_float32());
-            Array4f result_enoki = tex.eval_enoki(pos);
-            ek::eval(result_enoki);
-            Array4f result_cuda = tex.eval_cuda(pos);
-            ek::eval(result_cuda);
-
-            assert(ek::allclose(result_enoki, result_cuda, 5e-3f, 5e-3f));
+        Array<WrapMode, 3> wrap_modes(WrapMode::Repeat, WrapMode::Clamp,
+                                      WrapMode::Mirror);
```
Comment (Speierers) [inline tests/texture.cpp:91]: Looks like this could yield to lots of unnecessary compilation, e.g. introducing this new type `Array<WrapMode>`, maybe we could use `WrapMode wrap_modes[3] = { ... };` instead?

File: tests/texture.cpp [hunk 4]
```diff
@@ -78,41 +115,51 @@ ENOKI_TEST(test03_interp_2d) {
...
-        for (int i = 0; i < 4; ++i) {
-            tex.set_value(rng_1.next_float32());
-
-            Array2f pos(rng_2.next_float32(), rng_2.next_float32());
-            Array4f result_enoki = tex.eval_enoki(pos);
-            ek::eval(result_enoki);
-            Array4f result_cuda = tex.eval_cuda(pos);
-            ek::eval(result_cuda);
-            assert(ek::allclose(result_enoki, result_cuda, 5e-3f, 5e-3f));
+        Array<WrapMode, 3> wrap_modes(WrapMode::Repeat, WrapMode::Clamp,
```
Comment (Speierers) [inline tests/texture.cpp:118]: ditto

File: tests/texture.cpp [hunk 5]
```diff
@@ -78,41 +115,51 @@ ENOKI_TEST(test03_interp_2d) {
...
-            tex.set_value(rng_1.next_float32());
-
-            Array3f pos(rng_2.next_float32(), rng_2.next_float32(),
-                        rng_2.next_float32());
-            Array4f result_enoki = tex.eval_enoki(pos);
-            ek::eval(result_enoki);
-            Array4f result_cuda = tex.eval_cuda(pos);
-            ek::eval(result_cuda);
-            assert(ek::allclose(result_enoki, result_cuda, 5e-3f, 5e-3f));
+        Array<WrapMode, 3> wrap_modes(WrapMode::Repeat, WrapMode::Clamp,
```
Comment (Speierers) [inline tests/texture.cpp:145]: ditto

File: tests/texture.cpp [hunk 6]
```diff
@@ -179,38 +227,110 @@ ENOKI_TEST(test07_cubic_analytic) {
...
-    auto res = tex.eval_cubic(pos, true, true);
-    auto res2 = tex.eval_cubic_helper(pos);
-    assert(ek::allclose(res, res2, 1e-4f, 1e-4f));
-
-    PCG32<Float> rng(1024);
-    Array2f pos_(rng.next_float32(), rng.next_float32());
-    auto res_ = tex.eval_cubic(pos, true, true);
-    auto res2_ = tex.eval_cubic_helper(pos);
-    assert(ek::allclose(res_, res2_, 1e-4f, 1e-4f));
+    Array<WrapMode, 3> wrap_modes(WrapMode::Repeat, WrapMode::Clamp,
```
Comment (Speierers) [inline tests/texture.cpp:315]: ditto

File: include/enoki/texture.h [hunk 7]
```diff
@@ -10,29 +10,29 @@
...
 
 NAMESPACE_BEGIN(enoki)
 
-enum class FilterMode : uint32_t {
-    Nearest = 0,
-    Linear = 1
-};
+enum class FilterMode : uint32_t { Nearest = 0, Linear = 1 };
+
+enum class WrapMode : uint32_t { Repeat = 0, Clamp = 1, Mirror = 2 };
```
Comment (Speierers) [inline include/enoki/texture.h:26]: Let's extend those on multiple lines
```suggestion
enum class FilterMode : uint32_t { 
    Nearest = 0, 
    Linear = 1 
};

enum class WrapMode : uint32_t { 
    Repeat = 0, 
    Clamp = 1, 
    Mirror = 2 
};
```

Comment (Speierers) [reply include/enoki/texture.h:26]: Could also add short comments / descriptions

File: include/enoki/texture.h [hunk 8]
```diff
@@ -58,8 +58,9 @@ template <typename Value, size_t Dimension> class Texture {
      * operations need to perform a reverse migration in this mode.
```
Comment (Speierers) [inline include/enoki/texture.h:58]: Could mention the filter mode and warp mode in the docstring here. (and their default values)

File: include/enoki/texture.h [hunk 9]
```diff
@@ -72,24 +73,27 @@ template <typename Value, size_t Dimension> class Texture {
      * the texture memory with the provided tensor.
```
Comment (Speierers) [inline include/enoki/texture.h:73]: ditto

File: include/enoki/texture.h [hunk 10]
```diff
@@ -191,19 +198,39 @@ template <typename Value, size_t Dimension> class Texture {
...
-                                pos_idx, active.index(), out);
+            jit_cuda_tex_lookup(Dimension, m_handle_opaque.index(), pos_idx,
+                                active.index(), out);
 
-            return {
-                Value::steal(out[0]), Value::steal(out[1]),
-                Value::steal(out[2]), Value::steal(out[3])
-            };
+            return { Value::steal(out[0]), Value::steal(out[1]),
+                     Value::steal(out[2]), Value::steal(out[3]) };
```
Comment (Speierers) [inline include/enoki/texture.h:205]: Let's revert this change. 🙃 

Did you use clang-format for formatting? We are not super strict about formatting and don't always follow what comes out of clang-format btw.

File: include/enoki/texture.h [hunk 11]
```diff
@@ -191,19 +198,39 @@ template <typename Value, size_t Dimension> class Texture {
...
         } else {
             (void) pos; (void) active;
             return 0;
         }
     }
 
+    template <typename T> T wrap(const T &value) const {
+        const Array<Int32, Dimension> shape = m_shape_opaque;
```
Comment (Speierers) [inline include/enoki/texture.h:213]: Could add a static assertion to make sure `T` has the right dimension?

Comment (Speierers) [reply include/enoki/texture.h:213]: Also maybe this method should be private?

File: include/enoki/texture.h [hunk 12]
```diff
@@ -191,19 +198,39 @@ template <typename Value, size_t Dimension> class Texture {
...
+            T div;
+            for (size_t i = 0; i < Dimension; ++i)
+                div[i] = m_inv_resolution[i](value_shift_neg[i]);
+
+            T mod = value - div * shape;
+            masked(mod, mod < 0) += T(shape);
+
+            if (m_wrap_mode == WrapMode::Mirror)
+                mod =
+                    select(eq(div & 1, 0) ^ (value < 0), mod, shape - 1 - mod);
```
Comment (Speierers) [inline include/enoki/texture.h:228]: Complicated logic -> worth adding a comment here 😉

File: include/enoki/texture.h [hunk 13]
```diff
@@ -470,16 +526,18 @@ template <typename Value, size_t Dimension> class Texture {
             Value alpha2 = sqr(alpha),
                   alpha3 = alpha2 * alpha;
             Value multiplier = 1.f / 6.f;
-            // four basis functions, transformed to take as input the fractional part
-            Value w0 = (- alpha3 + 3.f * alpha2 - 3.f * alpha + 1.f) * multiplier,
+            // four basis functions, transformed to take as input the fractional
+            // part
```
Comment (Speierers) [inline include/enoki/texture.h:530]:
```suggestion
            // Four basis functions, transformed to take as input the fractional part
```

File: include/enoki/texture.h [hunk 14]
```diff
@@ -470,16 +526,18 @@ template <typename Value, size_t Dimension> class Texture {
...
+                  w3 = alpha3 * multiplier;
             Value w01 = w0 + w1,
                   w23 = 1.f - w01;
             return Array3(
-                w01,
-                (integ - 0.5f + w1 / w01) * inv_shape[dim],
-                (integ + 1.5f + w3 / w23) * inv_shape[dim]);  // (integ + 0.5) +- 1 + weight
+               w01,
+               (integ - 0.5f + w1 / w01) * inv_shape[dim],
+               (integ + 1.5f + w3 / w23) * inv_shape[dim]); // (integ + 0.5) +- 1 + weight
```
Comment (Speierers) [inline include/enoki/texture.h:540]: indentation?

File: include/enoki/texture.h [hunk 15]
```diff
@@ -535,76 +593,93 @@ template <typename Value, size_t Dimension> class Texture {
                replace the AD graph. The result is unused (and never computed)
                and only the AD graph is replaced. */
             if (grad_enabled(m_value, pos)) {
-                Array4 result_diff = eval_cubic_helper(pos, active);  // AD graph only
+                Array4 result_diff =
+                    eval_cubic_helper(pos, active); // AD graph only
```
Comment (Speierers) [inline include/enoki/texture.h:597]:
```suggestion
                Array4 result_diff = eval_cubic_helper(pos, active); // AD graph only
```

File: include/enoki/texture.h [hunk 16]
```diff
@@ -667,38 +742,44 @@ template <typename Value, size_t Dimension> class Texture {
...
     }
 
 private:
     void *m_handle = nullptr;
-    size_t m_shape[Dimension + 1] { };
     size_t m_size = 0;
     UInt64 m_handle_opaque;
     Array<UInt32, Dimension> m_shape_opaque;
     mutable TensorXf m_value;
+    Array<divisor<int32_t>, Dimension> m_inv_resolution;
```
Comment (Speierers) [inline include/enoki/texture.h:780]: Same here, maybe worth using a C-style array instead of this complicated enoki array?

File: include/enoki/texture.h [hunk 17]
```diff
@@ -213,80 +240,99 @@ template <typename Value, size_t Dimension> class Texture {
...
+                pos_i_w = wrap(InterpPosI(InterpIdx(0, 1) + pos_i.x()));
+            else if constexpr (Dimension == 2)
+                pos_i_w = wrap(InterpPosI(InterpIdx(0, 1, 0, 1) + pos_i.x(),
+                                          InterpIdx(0, 0, 1, 1) + pos_i.y()));
+            else if constexpr (Dimension == 3)
+                pos_i_w = wrap(
+                    InterpPosI(InterpIdx(0, 1, 0, 1, 0, 1, 0, 1) + pos_i.x(),
+                               InterpIdx(0, 0, 0, 0, 1, 1, 1, 1) + pos_i.y(),
+                               InterpIdx(0, 0, 1, 1, 0, 0, 1, 1) + pos_i.z()));
+
```
Comment (Speierers) [inline include/enoki/texture.h:287]:
```suggestion
            if constexpr (Dimension == 1) {
                pos_i_w = wrap(InterpPosI(InterpIdx(0, 1) + pos_i.x()));
            } else if constexpr (Dimension == 2) {
                pos_i_w = wrap(InterpPosI(InterpIdx(0, 1, 0, 1) + pos_i.x(),
                                          InterpIdx(0, 0, 1, 1) + pos_i.y()));
            } else if constexpr (Dimension == 3) {
                pos_i_w = wrap(
                    InterpPosI(InterpIdx(0, 1, 0, 1, 0, 1, 0, 1) + pos_i.x(),
                               InterpIdx(0, 0, 0, 0, 1, 1, 1, 1) + pos_i.y(),
                               InterpIdx(0, 0, 1, 1, 0, 0, 1, 1) + pos_i.z()));
            }

```

File: include/enoki/texture.h [hunk 18]
```diff
@@ -535,76 +593,93 @@ template <typename Value, size_t Dimension> class Texture {
...
                 return multiplier * Array4(
-                    -3.f * alpha2 +  6.f * alpha - 3.f,
-                     9.f * alpha2 - 12.f * alpha,
-                    -9.f * alpha2 +  6.f * alpha + 3.f,
-                     3.f * alpha2
-                );
+                    -3.f * alpha2 + 6.f * alpha - 3.f,
+                    9.f * alpha2 - 12.f * alpha,
+                    -9.f * alpha2 + 6.f * alpha + 3.f,
+                    3.f * alpha2);
```
Comment (Speierers) [inline include/enoki/texture.h:668]:
```suggestion
                    -3.f * alpha2 + 6.f * alpha - 3.f,
                     9.f * alpha2 - 12.f * alpha,
                    -9.f * alpha2 + 6.f * alpha + 3.f,
                     3.f * alpha2);
```

File: include/enoki/texture.h [hunk 19]
```diff
@@ -535,76 +593,93 @@ template <typename Value, size_t Dimension> class Texture {
...
+                                             bool is_grad) -> Array4 {
             const Value &alpha = pos_a[dim];
             Value alpha2 = alpha * alpha;
             Value multiplier = rcp(6.f);
             if (!is_grad) {
                 Value alpha3 = alpha2 * alpha;
                 return multiplier * Array4(
-                    - alpha3 + 3.f * alpha2 - 3.f * alpha + 1.f,
+                    -alpha3 + 3.f * alpha2 - 3.f * alpha + 1.f,
                     3.f * alpha3 - 6.f * alpha2 + 4.f,
```
Comment (Speierers) [inline include/enoki/texture.h:660]:
```suggestion
                     3.f * alpha3 - 6.f * alpha2 + 4.f,
```

File: include/enoki/texture.h [hunk 20]
```diff
@@ -392,29 +450,28 @@ template <typename Value, size_t Dimension> class Texture {
...
-                for (uint32_t iy=0; iy<4; iy++)
-                    EK_TEX_CUBIC_ACCUM(index + step[ix].x() + step[iy].y(), wx[ix] * wy[iy]);
+            Array4 wx = compute_weight(0), wy = compute_weight(1);
+            for (uint32_t ix = 0; ix < 4; ix++)
+                for (uint32_t iy = 0; iy < 4; iy++)
+                    EK_TEX_CUBIC_ACCUM(index[ix * 4 + iy], wx[ix] * wy[iy]);
         } else if constexpr (Dimension == 3) {
-            Array4 wx = compute_weight(0),
-                   wy = compute_weight(1),
+            Array4 wx = compute_weight(0), wy = compute_weight(1),
```
Comment (Speierers) [inline include/enoki/texture.h:468]:
```suggestion
            Array4 wx = compute_weight(0), 
                        wy = compute_weight(1),
```

File: include/enoki/texture.h [hunk 21]
```diff
@@ -337,53 +385,63 @@ template <typename Value, size_t Dimension> class Texture {
...
 
-        const uint32_t channels = (uint32_t) m_shape[Dimension];
+        InterpIdx index;
+        if constexpr (Dimension == 1)
+            index = pos_i_w.x();
+        else if constexpr (Dimension == 2)
+            index = fmadd(pos_i_w.y(), m_shape_opaque.x(), pos_i_w.x());
+        else if constexpr (Dimension == 3)
+            index = fmadd(fmadd(pos_i_w.z(), m_shape_opaque.y(), pos_i_w.y()),
+                          m_shape_opaque.x(), pos_i_w.x());
```
Comment (Speierers) [inline include/enoki/texture.h:441]: Maybe worth writing this as a for loop, especially given that the next task will be to support an arbitrary number of dimensions 😉 

This could also be part of the next PR of course.

File: include/enoki/texture.h [hunk 22]
```diff
@@ -337,53 +385,63 @@ template <typename Value, size_t Dimension> class Texture {
...
+
+        const auto compute_weight = [&pos_a](uint32_t dim) -> Array4 {
+            const Value &alpha = pos_a[dim];
+            Value alpha2 = alpha * alpha, alpha3 = alpha2 * alpha;
+            Value multiplier = rcp(6.f);
+            return multiplier *
+                   Array4(-alpha3 + 3.f * alpha2 - 3.f * alpha + 1.f,
+                          3.f * alpha3 - 6.f * alpha2 + 4.f,
+                          -3.f * alpha3 + 3.f * alpha2 + 3.f * alpha + 1.f,
+                          alpha3);
```
Comment (Speierers) [inline include/enoki/texture.h:431]:
```suggestion
                           3.f * alpha3 - 6.f * alpha2 + 4.f,
                          -3.f * alpha3 + 3.f * alpha2 + 3.f * alpha + 1.f,
                           alpha3);
```

File: include/enoki/texture.h [hunk 23]
```diff
@@ -337,53 +385,63 @@ template <typename Value, size_t Dimension> class Texture {
...
+                }
             }
         }
+        pos_i_w = wrap(pos_i_w);
+
+        PosF pos_a = pos_f - pos_i;
+
+        const auto compute_weight = [&pos_a](uint32_t dim) -> Array4 {
+            const Value &alpha = pos_a[dim];
+            Value alpha2 = alpha * alpha, alpha3 = alpha2 * alpha;
```
Comment (Speierers) [inline include/enoki/texture.h:425]: We tend to avoid multiple declaration on a single line.
```suggestion
            Value alpha2 = alpha * alpha, 
                      alpha3 = alpha2 * alpha;
```

File: include/enoki/texture.h [hunk 24]
```diff
@@ -337,53 +385,63 @@ template <typename Value, size_t Dimension> class Texture {
...
+                            offset[iz] + pos_i.x();
+                        pos_i_w[1][ix * 16 + iz * 4 + iy] =
+                            offset[iz] + pos_i.y();
+                        pos_i_w[2][iy * 16 + ix * 4 + iz] =
+                            offset[iz] + pos_i.z();
+                    }
+                }
             }
         }
+        pos_i_w = wrap(pos_i_w);
```
Comment (Speierers) [inline include/enoki/texture.h:419]:
```suggestion
        }
        
        pos_i_w = wrap(pos_i_w);
```

File: include/enoki/texture.h [hunk 25]
```diff
@@ -213,80 +240,99 @@ template <typename Value, size_t Dimension> class Texture {
...
+
+            UInt32 index;
+            if constexpr (Dimension == 1) {
+                index = pos_i_w.x();
+            } else if constexpr (Dimension == 2) {
+                index = fmadd(pos_i_w.y(), m_shape_opaque.x(), pos_i_w.x());
+            } else if constexpr (Dimension == 3) {
+                index =
+                    fmadd(fmadd(pos_i_w.z(), m_shape_opaque.y(), pos_i_w.y()),
+                          m_shape_opaque.x(), pos_i_w.x());
```
Comment (Speierers) [inline include/enoki/texture.h:260]: Those few lines seems to come back a few times accross this file, could be worth moving it in a private method.

**Review by Speierers — COMMENTED** [review]
Here is a first pass over this PR.

Please when you address the comments, mark them as resolved so it is easier to keep track of what's left to be done in the PR.

Good job overall!

File: include/enoki/texture.h [hunk 26]
```diff
@@ -58,8 +58,9 @@ template <typename Value, size_t Dimension> class Texture {
      * operations need to perform a reverse migration in this mode.
      */
     Texture(const size_t shape[Dimension], size_t channels, bool migrate = true,
-            FilterMode filter_mode = FilterMode::Linear) {
-        init(shape, channels, migrate, filter_mode);
+            FilterMode filter_mode = FilterMode::Linear,
+            WrapMode wrap_mode = WrapMode::Repeat) {
```
Comment (wjakob) [inline include/enoki/texture.h:62]: Let's have Clamp as the default.

File: include/enoki/texture.h (hunk 25 above)
Comment (wjakob) [reply include/enoki/texture.h:260]: Agreed, maybe just `UInt32 index = index(pos_i_w)`. Note that this `index()` function should also do an explicit signed-unsigned int cast.

```
UInt32 index(const Array<Value, Int32> &pos_i) {
    using PosU = Array<Value, Unt32>;
    PosU pos_u(pos_i);
    ...
}
```

File: include/enoki/texture.h [hunk 27]
```diff
@@ -213,80 +240,99 @@ template <typename Value, size_t Dimension> class Texture {
...
             for (uint32_t ch = 0; ch < channels; ++ch)
                 result[ch] = gather<Value>(m_value.array(), index + ch, active);
+
             return result;
-        }
+        } else {
+            using InterpIdx = Array<Int32, 1 << Dimension>;
+            using InterpPosI = Array<InterpIdx, Dimension>;
+
+            const PosF pos_f = fmadd(pos, m_shape_opaque, -.5f);
```
Comment (wjakob) [inline include/enoki/texture.h:273]: make cast explicit, i.e. `PosF(m_shape_opaque)`?

File: include/enoki/texture.h [hunk 28]
```diff
@@ -213,80 +240,99 @@ template <typename Value, size_t Dimension> class Texture {
...
+
+            InterpPosI pos_i_w;
+            if constexpr (Dimension == 1)
+                pos_i_w = wrap(InterpPosI(InterpIdx(0, 1) + pos_i.x()));
+            else if constexpr (Dimension == 2)
+                pos_i_w = wrap(InterpPosI(InterpIdx(0, 1, 0, 1) + pos_i.x(),
+                                          InterpIdx(0, 0, 1, 1) + pos_i.y()));
+            else if constexpr (Dimension == 3)
+                pos_i_w = wrap(
+                    InterpPosI(InterpIdx(0, 1, 0, 1, 0, 1, 0, 1) + pos_i.x(),
```
Comment (wjakob) [inline include/enoki/texture.h:284]: Great! A lot of this computation is redundant, but Enoki's common subexpression elimination should be able to remove any duplicated arithmetic.

File: include/enoki/texture.h [hunk 29]
```diff
@@ -317,7 +363,7 @@ template <typename Value, size_t Dimension> class Texture {
     }
 
     /**
-     * \brief Helper function to evaluate a clamped cubic B-Spline interpolant
```
Comment (wjakob) [inline include/enoki/texture.h:320]: Note that "clamped" here refers to a property of the spline basis functions rather than the texture wrapping mode :D

Comment (wjakob) [reply include/enoki/texture.h:320]: I am impressed that you also made this work for the b-spline interpolation variant ;-) (I had somehow assumed we'd give up on this for now and always assume clamped wrapping mode here -- anyways, all the better)

File: include/enoki/texture.h (hunk 21 above)
Comment (wjakob) [reply include/enoki/texture.h:441]: Small correction: we want to support an arbitrary number of channels, but the dimension will be restricted to 1/2/3D.

File: include/enoki/texture.h [hunk 30]
```diff
@@ -337,53 +385,63 @@ template <typename Value, size_t Dimension> class Texture {
...
+                            offset[iz] + pos_i.y();
+                        pos_i_w[2][iy * 16 + ix * 4 + iz] =
+                            offset[iz] + pos_i.z();
+                    }
+                }
             }
         }
+        pos_i_w = wrap(pos_i_w);
+
+        PosF pos_a = pos_f - pos_i;
```
Comment (wjakob) [inline include/enoki/texture.h:421]: Explicit cast: `PosF(pos_i)`

File: include/enoki/texture.h [hunk 31]
```diff
@@ -337,53 +385,63 @@ template <typename Value, size_t Dimension> class Texture {
...
+
+        const auto compute_weight = [&pos_a](uint32_t dim) -> Array4 {
+            const Value &alpha = pos_a[dim];
+            Value alpha2 = alpha * alpha, alpha3 = alpha2 * alpha;
+            Value multiplier = rcp(6.f);
```
Comment (wjakob) [inline include/enoki/texture.h:426]: using (generally aproximate) `rcp` is overkill for scalar constants: you can write `multiplier = 1.f / 6.f`.

File: tests/texture.cpp (hunk 3 above)
Comment (wjakob) [reply tests/texture.cpp:91]: Seems fine to me, it's just for the test suite in any case.

**Review by wjakob — COMMENTED** [review]
Amazing work @njroussel  -- I just had a few small comments. Please feel free to merge once you've had a look at the feedback from @Speierers and me.

**Comment by wjakob** [comment]
By the way, feel free to add the names of yourself and Ziyi to the copyright header at the top. You did just at least as much of the work as me.

**Comment by wjakob** [comment]
Is it possible that this isn't fully rebased on master? I had at some point done a commenting and touchup pass that doesn't seem reflected here, see 

- https://github.com/wjakob/enoki/commit/5f522c92339a90f34d7c845ac779eda02fad4fc7#diff-11e198addd80048df5aaba783cdc3e6591ecc76ec2b917b6961f0c27feb9baf6

- https://github.com/wjakob/enoki/commit/c653ad5c4b6b8b78f47daa4f74c4d45ade88a3ea#diff-11e198addd80048df5aaba783cdc3e6591ecc76ec2b917b6961f0c27feb9baf6

File: include/enoki/texture.h (hunk 10 above)
Comment (njroussel) [reply include/enoki/texture.h]: Yes, I was applying clang-format without second-guessing myself - I thought Enoki shared the same configuration as Mitsuba which provides a `.clang-format` file.

I removed a lot of these changes so that the diff would be more readable for the PR, but I clearly missed a few. I'll be more careful from now on :smiley:

File: include/enoki/texture.h (hunk 11 above)
Comment (njroussel) [reply include/enoki/texture.h]: Totally makes sense, I changed it to a private method with the proper type checks (signed input of the correct length).

File: include/enoki/texture.h (hunk 25 above)
Comment (njroussel) [reply include/enoki/texture.h]: I moved this whole logic to a templated private method, we now have the same method to compute indices for linear interpolation and clamped b-spline interpolation.

Something very similar could be done for the computation of all the input points that participate in the interpolant, I did it too.

File: include/enoki/texture.h [hunk 32]
```diff
@@ -785,28 +741,54 @@ template <typename Value, size_t Dimension> class Texture {
...
+                std::is_integral_v<Scalar> &&
+                std::is_signed_v<Scalar>
+        );
+
+        Index index;
+        if constexpr (Dimension == 1)
+            index = Index(pos.x());
+        else if constexpr (Dimension == 2)
+            index =
+                Index(fmadd(pos.y(), (Int32) m_shape_opaque.x(), pos.x()));
```
Comment (wjakob) [inline include/enoki/texture.h:781]: Rather than casting `m_shape_opaque` to a signed integer, can you cast the positions to unsigned ones? (the input positions are already wrapped/clamped, so should be >=0).

Context: we've actually run into 32 bit indexing overflows several times with Mitsuba before (users do crazy things) -- here, we are already committed to 32 bit, so let's at least not give up the top bit to sign information ;-).

Comment (njroussel) [reply include/enoki/texture.h]: Indeed! I hadn't been paying much attention to these type promotions so far :sweat_smile: 
Should be good now

**Comment by njroussel** [comment]
@wjakob Regarding your comment:

> Is it possible that this isn't fully rebased on master? I had at some point done a commenting and touchup pass that doesn't seem reflected here, see
> 
>     * [5f522c9#diff-11e198addd80048df5aaba783cdc3e6591ecc76ec2b917b6961f0c27feb9baf6](https://github.com/wjakob/enoki/commit/5f522c92339a90f34d7c845ac779eda02fad4fc7#diff-11e198addd80048df5aaba783cdc3e6591ecc76ec2b917b6961f0c27feb9baf6)
> 
>     * [c653ad5#diff-11e198addd80048df5aaba783cdc3e6591ecc76ec2b917b6961f0c27feb9baf6](https://github.com/wjakob/enoki/commit/c653ad5c4b6b8b78f47daa4f74c4d45ade88a3ea#diff-11e198addd80048df5aaba783cdc3e6591ecc76ec2b917b6961f0c27feb9baf6)

Unless I'm  missing something obivous, those changes are all currently included on this branch. You might have gotten slightly confused, as I did, as some of these lines/fixes should have been duplicated, like here:https://github.com/wjakob/enoki/blob/c653ad5c4b6b8b78f47daa4f74c4d45ade88a3ea/include/enoki/texture.h#L357

**Comment by wjakob** [comment]
@njroussel -- I likely got confused then :). Please feel free to merge once you're done here (let's squash this into a single commit though).

**Comment by njroussel** [comment]
@Speierers I resolved all the comments you initially had. Did you want to have a second pass over the code now ?

@wjakob Looks like I don't have write access to this repository, I can't merge pull requests :slightly_frowning_face:

**Comment by wjakob** [comment]
added you.

File: include/enoki/texture.h [hunk 33]
```diff
@@ -667,38 +652,150 @@ template <typename Value, size_t Dimension> class Texture {
...
-                                           (int) filter_mode);
+                                           (int) filter_mode, (int) wrap_mode);
             m_handle_opaque = UInt64::steal(
                 jit_var_new_pointer(JitBackend::CUDA, m_handle, 0, 0));
         }
     }
 
 private:
+    template <typename T>
+        constexpr static T ipow(T num, unsigned int pow) {
```
Comment (Speierers) [inline include/enoki/texture.h:686]: Indentation

File: include/enoki/texture.h [hunk 34]
```diff
@@ -667,38 +652,150 @@ template <typename Value, size_t Dimension> class Texture {
...
+    template <typename T, size_t Length>
+    Array<Array<Int32, ipow(Length, Dimension)>, Dimension>
+    interp_positions(const int *offset, const T &pos) const {
+        using Scalar = scalar_t<T>;
+        using InterpOffset = Array<Int32, ipow(Length, Dimension)>;
+        using InterpPosI = Array<InterpOffset, Dimension>;
+        static_assert(
+                array_size_v<T> == Dimension &&
+                std::is_integral_v<Scalar> &&
+                std::is_signed_v<Scalar>
```
Comment (Speierers) [inline include/enoki/texture.h:699]: Indentation

File: include/enoki/texture.h [hunk 35]
```diff
@@ -667,38 +652,150 @@ template <typename Value, size_t Dimension> class Texture {
...
+
+        return pos_i;
+    }
+
+    template <typename T> T wrap(const T &pos) const {
+        using Scalar = scalar_t<T>;
+        static_assert(
+                array_size_v<T> == Dimension &&
+                std::is_integral_v<Scalar> &&
+                std::is_signed_v<Scalar>
```
Comment (Speierers) [inline include/enoki/texture.h:738]: Indentation

File: include/enoki/texture.h [hunk 36]
```diff
@@ -667,38 +652,150 @@ template <typename Value, size_t Dimension> class Texture {
...
+    }
+
+    template <typename T>
+    uint32_array_t<value_t<T>> index(const T &pos) const {
+        using Scalar = scalar_t<T>;
+        using Index = uint32_array_t<value_t<T>>;
+        static_assert(
+                array_size_v<T> == Dimension &&
+                std::is_integral_v<Scalar> &&
+                std::is_signed_v<Scalar>
```
Comment (Speierers) [inline include/enoki/texture.h:772]: ditto

File: include/enoki/texture.h [hunk 37]
```diff
@@ -667,38 +652,150 @@ template <typename Value, size_t Dimension> class Texture {
...
+        Index index;
+        if constexpr (Dimension == 1)
+            index = Index(pos.x());
+        else if constexpr (Dimension == 2)
+            index = Index(
+                fmadd(Index(pos.y()), m_shape_opaque.x(), Index(pos.x())));
+        else if constexpr (Dimension == 3)
+            index = Index(fmadd(
+                fmadd(Index(pos.z()), m_shape_opaque.y(), Index(pos.y())),
+                m_shape_opaque.x(), Index(pos.x())));
```
Comment (Speierers) [inline include/enoki/texture.h:784]: I personnaly find it more readible with the braces ;)
```suggestion
        if constexpr (Dimension == 1) {
            index = Index(pos.x());
        } else if constexpr (Dimension == 2) {
            index = Index(
                fmadd(Index(pos.y()), m_shape_opaque.x(), Index(pos.x())));
        } else if constexpr (Dimension == 3) {
            index = Index(fmadd(
                fmadd(Index(pos.z()), m_shape_opaque.y(), Index(pos.y())),
                m_shape_opaque.x(), Index(pos.x())));
         }
```

**Review by Speierers — COMMENTED** [review]
Just a couple minor indentation issues, otherwise it looks good to me! You can merge this one! 🚀 Great job!
